feat: OCA provider

nihar-oracle opened the pull request:
feat: OCA provider

Oracle Code Assist provides a lite llm compatible endpoint for accessing LLM's. The following PR is to add OCA as a provider to cline.

We are basing our changes off of existing litellm provider implementation for field names, handlers etc.

<!--
Thank you for contributing to Cline!

⚠️ Important: Before submitting this PR, please ensure you have:
- Received approval from a core Cline contributor prior to proceeding with the implementation:

Limited exceptions:
Small bug fixes, typo corrections, minor wording improvements, or simple type fixes that don't change functionality may be submitted directly.

Why this requirement?
We are adding Oracle Code Assist as a provider to be adopted.
-->

### Related Issue

<!-- Replace XXXX with the issue number that this PR addresses -->
**Issue:** [#4986](https://github.com/cline/cline/discussions/4986#discussioncomment-13808844)

### Description

<!-- 
Help reviewers understand your changes by making this PR readable and well-organized:

- What problem does this PR solve?
Add Oracle Code Assist (oca) AI as a provider with necessary API, configuration, and UI updates.

- Why were these changes introduced and what purpose do they serve?
These changes were made to make oracle code assist available as a provider

- For larger changes, provide context about your approach and reasoning
* Behavior:
- Oracle Code Assist (OCA) is implemented via SSO / Oauth with an IDCS provider for ouath.
- Adds oca as a new API provider in proto/models.proto and proto/state.proto.
- Adds oca model refresh in proto/models.
- Adds oca service in proto/oca for login, logout and refresh.
- Implements Ocahandler (Extension of LiteLlmHandler) in src/api/providers/moonshot.ts to handle API interactions.
- Updates createHandlerForProvider() in src/api/index.ts to include Ocahandler.

* Configuration:
- Adds ocaAccessKey and other necessary fields to ApiConfiguration in src/shared/api.ts and src/core/storage/state.ts.
- Updates convertApiConfigurationToProto() and convertProtoToApiConfiguration() in src/shared/proto-conversions/models/api-configuration-conversion.ts to handle oca provider fields.

* UI:
- Adds OcaProvider component in webview-ui/src/components/settings/providers/OcaProvider.tsx along with OcaModelPicker.tsx component.
- Updates ApiOptions in webview-ui/src/components/settings/ApiOptions.tsx to include oca in the provider dropdown.
- Validates ocaAccessKey in webview-ui/src/utils/validate.ts.

-->

### Test Procedure

<!-- 
Please walk us through your testing approach and thought process. This helps reviewers understand that you've thoroughly considered the impact of your changes:

- How did you test this change?
- What could potentially break and how did you verify it doesn't?
- What existing functionality might be affected and how did you check it still works?
- Why are you confident this is ready for merge?

We're not looking for exhaustive documentation - just evidence that you've thought through the implications of your changes and tested accordingly.
-->

### Type of Change

<!-- Put an 'x' in all boxes that apply -->

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [x] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

<!-- Put an 'x' in all boxes that apply -->

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [x] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

<!-- 
Help reviewers quickly understand your changes:

- **UI Changes**: Please include screenshots showing before/after states
<img width="388" height="872" alt="image" src="https://github.com/user-attachments/assets/f320f84f-c277-4779-9219-ec43eab2db03" />


This helps reviewers see what you've built without having to pull down and test your branch first.
-->

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Adds Oracle Code Assist (OCA) as a new provider with API, configuration, and UI updates, including new models, handlers, and UI components.
> 
>   - **Behavior**:
>     - Adds Oracle Code Assist (OCA) as a new provider with SSO/OAuth via IDCS.
>     - Implements `OcaHandler` in `src/api/providers/oca.ts` for API interactions.
>     - Updates `createHandlerForProvider()` in `src/api/index.ts` to include `OcaHandler`.
>   - **Models**:
>     - Adds `OcaLiteLLMModelInfo` and `OcaModelInfoMap` to `proto/models.proto`.
>     - Adds `oca.proto` for OCA service definitions.
>   - **Configuration**:
>     - Adds OCA fields to `ApiConfiguration` in `src/shared/api.ts`.
>     - Updates `convertApiConfigurationToProto()` and `convertProtoToApiConfiguration()` in `proto-conversions/models/api-configuration-conversion.ts`.
>   - **UI**:
>     - Adds `OcaProvider` and `OcaModelPicker` components in `webview-ui/src/components/settings/providers/`.
>     - Updates `ApiOptions` in `webview-ui/src/components/settings/ApiOptions.tsx` to include OCA.
>   - **Misc**:
>     - Adds `OcaServiceClient` to `grpc-client.js`.
>     - Updates `ExtensionStateContext` in `webview-ui/src/context/ExtensionStateContext.tsx` to manage OCA state.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for a61635554f1aa403651a4ccf5d3cfb70183a7797. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

celestial-vault commented on proto/models.proto:
```diff
@@ -229,6 +254,10 @@ message ModelsApiConfiguration {
   optional string cline_account_id = 58;
   optional string groq_api_key = 59;
   optional string hugging_face_api_key = 60;
+  optional string oca_lite_llm_base_url = 61;
```
why is is called oca_lite_llm? Is it OCA or LiteLLM?

celestial-vault commented on src/api/providers/litellm.ts:
```diff
@@ -6,7 +6,7 @@ import { ApiStream } from "../transform/stream"
 import { convertToOpenAiMessages } from "../transform/openai-format"
 import { withRetry } from "../retry"
```
Could you write in the description of this PR the connection between the Oracle provider and LightLLM? I'm confused as to why we're adding all of these LiteLLM fields and changing this seemingly unconnected provider.

nihar-oracle replied:
Of course, I will update the description.
So, it is OCA's litellm compatible endpoint. The only real difference between this and a proper litellm implementation is that oca uses sso oauth backed access keys instead of api keys.

I am happy to rename this to oca_ instead of oca_litellm if that is better.
I have also created the oca handler as a derived class of litellm handler, I am happy to change that to it's own independent class as well

celestial-vault replied:
Yeah let's rename it

celestial-vault commented on src/api/providers/oca.ts:
```diff
@@ -0,0 +1,63 @@
+import { DEFAULT_OCA_BASE_URL } from "@/core/controller/oca/util/constants"
+import { createOcaHeaders } from "@/core/controller/oca/util/utils"
+
+export class OcaHandler extends LiteLlmHandler {
```
If this is somehow built on top of LiteLLM, I would still recommend writing it separately and not tying it to the LiteLLM handler code.

celestial-vault replied:
Yeah, let's move it into it's own class so they don't get mixed up.

celestial-vault commented on src/api/providers/oca.ts:
```diff
@@ -0,0 +1,63 @@
+						throw new Error("Oracle Code Assist (OCA) access token is not available")
+					}
+
+					const globalFetch = (typeof fetch === "function" ? fetch : globalThis.fetch).bind(undefined)
```
If you were to do that, you would also avoid complexity here.

nihar-oracle replied:
Happy to move this to it's own class, just that the only difference between the two classes is the auth implementation.

nihar-oracle replied:
done

celestial-vault commented on src/core/controller/web/openInBrowser.ts:
```diff
@@ -8,7 +8,7 @@ import { openExternal } from "@utils/env"
  * @returns Empty response since the client doesn't need a return value
  */
-export async function openInBrowser(controller: Controller, request: StringRequest): Promise<Empty> {
+export async function openInBrowser(controller: Controller | null, request: StringRequest): Promise<Empty> {
```
Why would the controller be null?

celestial-vault commented on webview-ui/src/components/settings/OcaModelPicker.tsx:
```diff
@@ -0,0 +1,248 @@
+	bannerText?: string | null
+}> = React.memo(({ onAcknowledge, bannerText }) => (
+	<div
+		style={{
```
Let's use Tailwind here

celestial-vault reviewed (commented):
Looking good!

Let's make some changes based on the above comments.

As to the authentication, can we get it consolidated as much as possible?

nihar-oracle commented on webview-ui/src/context/ExtensionStateContext.tsx:
```diff
@@ -634,6 +645,39 @@ export const ExtensionStateContextProvider: React.FC<{
 			.catch((error: Error) => console.error("Failed to refresh OpenRouter models:", error))
 	}, [])
 
+	const refreshOcaModels = useCallback((url: string) => {
+		ModelsServiceClient.refreshOcaModels(StringRequest.create({ value: url }))
+			.then((response: OcaModelInfoMap) => {
+				const models = response.models
+				setOcaModels(models)
+			})
+			.catch((error: Error) => console.error("Failed to refresh Oca models:", error))
+	}, [])
+
+	const ocaRefreshToken = useCallback(() => {
+		OcaServiceClient.ocaRefreshToken(EmptyRequest.create({}))
+			.then(() => {
+				console.log("OCA token refreshed successfully")
+			})
+			.catch((error: Error) => console.error("Failed to refresh OCA token:", error))
+	}, [])
+
+	const ocaLogin = useCallback(() => {
+		OcaServiceClient.ocaLoginClicked(EmptyRequest.create({}))
+			.then(() => {
+				console.log("OCA login successful")
+			})
+			.catch((error: Error) => console.error("Failed to login to OCA:", error))
+	}, [])
+
+	const ocaLogout = useCallback(() => {
+		OcaServiceClient.ocaLogoutClicked(EmptyRequest.create({}))
+			.then(() => {
+				console.log("OCA logout successful")
+			})
+			.catch((error: Error) => console.error("Failed to logout from OCA:", error))
+	}, [])
+
```
Of course.
I thought this was the pattern that had to be followed and I borrowed it from the other structures in the extension context. Does not need to be in there at all. But the models are refreshed and set on each mount.

Whatever is recommended

nihar-oracle commented:
@celestial-vault just checking in about some of the questions in my response

kefayati commented:
Thank you @celestial-vault for feedback. Your response on some of the above is much appreciated. Our users are eager to start using main branch Cline with OCA.

celestial-vault commented on package.json:
```diff
@@ -457,6 +458,7 @@
 		"diff": "^5.2.0",
 		"exceljs": "^4.4.0",
 		"execa": "^9.5.2",
+		"express": "^5.1.0",
```
Are these package.json changes absolutely necessary for this provider?

nihar-oracle replied:
We need to spin up a local host server to complete the oauth unfortunately, happy to tie it into vscode the way cline does it, but wanted to keep things separate

nihar-oracle replied:
Open to suggestions

celestial-vault commented on proto/oca.proto:
```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
```
sso is indeed unique

celestial-vault commented on src/core/controller/oca/util/ocaTokenManager.ts:
```diff
@@ -0,0 +1,227 @@
+			expires_at: expires_at!,
+			sub,
+		}
+		await fs.writeFile(this.TOKEN_CACHE_PATH, JSON.stringify(toStore), { encoding: "utf-8" })
```
Can this not just be stored in the VS Code secrets so we don't need to do this chmod?

celestial-vault replied:
We do store it in the vscode secrets, but refreshing the secrets through the api handlers is currently not supported.

Sorry, not sure what you mean here. See my comment below. I don't see why the key can't be stored and read from VSCode secrets. Refreshing the secret would just be setting it again, no?

nihar-oracle replied:
No, to update the secrets with the refreshed value of the token. Open to suggestions, but it will be a bigger change and and does get a little channeling when we convert between proto and non proto versions of the api config

celestial-vault commented on package.json:
```diff
@@ -361,8 +361,8 @@
-		"test:e2e": "playwright install && vsce package --no-dependencies --out dist/e2e.vsix && node src/test/e2e/utils/build.js && playwright test",
-		"test:e2e:optimal": "vsce package --no-dependencies --out dist/e2e.vsix && node src/test/e2e/utils/build.js && playwright test",
+		"test:e2e": "playwright install && vsce package --no-dependencies --allow-package-secrets sendgrid --out dist/e2e.vsix && node src/test/e2e/utils/build.js && playwright test",
+		"test:e2e:optimal": "vsce package --no-dependencies --allow-package-secrets sendgrid --out dist/e2e.vsix && node src/test/e2e/utils/build.js && playwright test",
```
How come we're changing the e2e tests here?

celestial-vault reviewed (commented):
Frankly speaking, this authentication stuff is going to take a bit to review. You seem to be adding in a lot of logic here, for example, setting up an express server just to listen for the response.

I realize that Cline needs authentication methods for API Providers other than just API keys, and you will even see that we've added in custom logic for our own provider for OAuth, but this needs to be standardized in some way. We just can't have everyone shoving in a bunch of logic for their own methods - it's not sustainable, and I'm concerned about the side-effects.

Is there no way that this can be simplified for the time being? Moving forward, we can set up a communication channel to discuss prioritizing some of these other methods. Potentially you can look to the SAP Provider for inspiration.

kefayati commented:
> Is there no way that this can be simplified for the time being? Moving forward, we can set up a communication channel to discuss prioritizing some of these other methods. Potentially you can look to the SAP Provider for inspiration.

We have made an effort to make it as simple as possible. We have looked into SAP's implementation as well. Happy to establish the long term communication channel since our team is investing code assist and code agents for Oracle users and customers alike. Feel free to let us know what is your preferred method of communication and I get it going.

nihar-oracle commented:
> Frankly speaking, this authentication stuff is going to take a bit to review. You seem to be adding in a lot of logic here, for example, setting up an express server just to listen for the response.
> 
> I realize that Cline needs authentication methods for API Providers other than just API keys, and you will even see that we've added in custom logic for our own provider for OAuth, but this needs to be standardized in some way. We just can't have everyone shoving in a bunch of logic for their own methods - it's not sustainable, and I'm concerned about the side-effects.
> 
> Is there no way that this can be simplified for the time being? Moving forward, we can set up a communication channel to discuss prioritizing some of these other methods. Potentially you can look to the SAP Provider for inspiration.

Thank you @celestial-vault .

I am open to having a discussion about this. I reached out on the discord channel as well.
Perhaps we can generalize this approach for other providers as well?

Anyone who needs to use oauth and access tokens?

Happy to add that feature in with this PR

celestial-vault commented:
> > Is there no way that this can be simplified for the time being? Moving forward, we can set up a communication channel to discuss prioritizing some of these other methods. Potentially you can look to the SAP Provider for inspiration.
> 
> We have made an effort to make it as simple as possible. We have looked into SAP's implementation as well. Happy to establish the long term communication channel since our team is investing code assist and code agents for Oracle users and customers alike. Feel free to let us know what is your preferred method of communication and I get it going. I am fairly visible in LinkedIn and we can get started from there.

Connected with you there

nihar-oracle commented:
> > > Is there no way that this can be simplified for the time being? Moving forward, we can set up a communication channel to discuss prioritizing some of these other methods. Potentially you can look to the SAP Provider for inspiration.
> > 
> > 
> > We have made an effort to make it as simple as possible. We have looked into SAP's implementation as well. Happy to establish the long term communication channel since our team is investing code assist and code agents for Oracle users and customers alike. Feel free to let us know what is your preferred method of communication and I get it going. I am fairly visible in LinkedIn and we can get started from there.
> 
> Connected with you there

Thank you, can you please message me there? I do not think I got the invite?

celestial-vault commented:
> > Is there no way that this can be simplified for the time being? Moving forward, we can set up a communication channel to discuss prioritizing some of these other methods. Potentially you can look to the SAP Provider for inspiration.
> 
> We have made an effort to make it as simple as possible. We have looked into SAP's implementation as well. Happy to establish the long term communication channel since our team is investing code assist and code agents for Oracle users and customers alike. Feel free to let us know what is your preferred method of communication and I get it going. I am fairly visible in LinkedIn and we can get started from there.

@kefayati Added you

nihar-oracle commented:
> Following up on the question above:
> 
> The current implementation stores tokens in with Unix file permissions (0600). This approach has several issues:
> 
> 1. **Cross-platform compatibility**: The permission check  is ineffective on Windows. While the code will run, Windows uses ACLs instead of Unix permissions, so the security validation () won't provide the intended protection on Windows systems.
> 2. **Security**: VSCode provides the SecretStorage API specifically for storing sensitive data like tokens. It handles encryption and secure storage across all platforms automatically.
> 3. **VSCode extension guidelines**: File-based token storage violates VSCode's security best practices for extensions.
> 
> Looking at the code:
> 
> ```ts
> private static readonly TOKEN_CACHE_PATH = path.join(os.homedir(), ".oca", "token_cache.json")
> 
> // Later in loadToken()
> const stats = await fs.stat(this.TOKEN_CACHE_PATH)
> const mode = stats.mode & 0o777
> if (mode !== 0o600) {
>     fs.rm(this.TOKEN_CACHE_PATH)
>     throw new Error("Token file permissions should be owner read and write")
> }
> ```
> 
> Could you clarify:
> 
> 1. Is there a specific requirement from Oracle that prevents using VSCode's SecretStorage API?
> 2. Do these tokens need to be shared with other Oracle CLI tools outside of VSCode?
> 3. Would you be open to using VSCode's SecretStorage as the primary storage mechanism, with file-based storage as an optional fallback for specific use cases?
> 
> This would make the extension more secure and work properly across all platforms that VSCode supports.

Thank you.
The windows check will be corrected, maybe I can research a cross platform way of checking.

To answer your other question, while we do intend to use this token across different integrations of the OCA, thereby reducing the number of sign-ins, the real goal for the introduction of the file is to refresh the token during the api call.

Currently, api handlers do not get the extension context passed down to them, so they cannot update vscode secrets.
Perhaps, we can discuss an alternate solution for the refresh mechanism that would not require external caching?

1. We can pass in the extension context to the api handlers to refresh teh secrets.
2. Set up a refresh mechanism in the index.ts of the controller to run a cron job that can, among other things, performa refreshes async.

Open to suggestions

nihar-oracle commented on src/core/controller/oca/util/ocaTokenManager.ts:
```diff
@@ -0,0 +1,227 @@
+		const config = await this.getConfig()
+		for (const port of config.port_candidates) {
+			try {
+				const server = http.createServer(() => {})
+				await new Promise((resolve, reject) => {
+					server.once("error", reject)
+					server.listen(port, "127.0.0.1", () => {
+						server.close()
+						resolve(true)
+					})
+				})
+				return port
```
So, I am open to switching to a vscode uri, which I agree is better.
To your point, I agree there is a race condition, fixing it by checking and instantly binding

celestial-vault replied:
Yeah, as discussed let's switch to a vscode uri approach

celestial-vault replied:
Let's do whatever's simplest for now, with the intent of being iterative here towards making a more general endpoint that we can all share

nihar-oracle committed:
feat(oca): add Oracle Code Assist provider; auth, models, settings

- Implement OCA API handler (src/core/api/providers/oca.ts)

  - OpenAI-compatible chat.completions with custom fetch injecting OCA headers
  - Optional reasoning (thinking) budget and ephemeral prompt caching
  - LiteLLM session tracking and usage streaming; cost via /spend/calculate
  - Guarded client init; clear error when OCA access token is missing

- Wire provider into core API and controller

  - Register provider (src/core/api/index.ts)
  - Controller flows for OCA account login/logout and auth status subscription
  - refreshOcaModels command and model config plumbing

- Add protobuf surfaces

  - proto/cline/models.proto and proto/cline/ocaAccount.proto
  - Extend proto/cline/state.proto for settings/state

- Persist settings/state and helpers
  - Update StateManager, state-keys, state-helpers, updateSettings

- Misc

  - Add changeset entry
  - Minor .gitignore and commit-message generator tweak

feat: New redirect server

feat(oca): add Oracle Code Assist provider; auth, models, settings

- Implement OCA API handler (src/core/api/providers/oca.ts)

  - OpenAI-compatible chat.completions with custom fetch injecting OCA headers
  - Optional reasoning (thinking) budget and ephemeral prompt caching
  - LiteLLM session tracking and usage streaming; cost via /spend/calculate
  - Guarded client init; clear error when OCA access token is missing

- Wire provider into core API and controller

  - Register provider (src/core/api/index.ts)
  - Controller flows for OCA account login/logout and auth status subscription
  - refreshOcaModels command and model config plumbing

- Add protobuf surfaces

  - proto/cline/models.proto and proto/cline/ocaAccount.proto
  - Extend proto/cline/state.proto for settings/state

- Persist settings/state and helpers
  - Update StateManager, state-keys, state-helpers, updateSettings

- Misc

  - Add changeset entry
  - Minor .gitignore and commit-message generator tweak

feat: New redirect server

update UI and add NPS survey link

papercuts

fix model dropdown height

fix: Fix 1

fix: removing ui

feat(AuthManager): Adding an AuthManager

fix: fixing rebase

nihar-oracle commented on src/integrations/git/commit-message-generator.ts:
```diff
@@ -78,7 +79,7 @@ Commit message:`
 
 		// Build the API handler
-		const apiHandler = buildApiHandler(apiConfiguration, currentMode)
+		const apiHandler = buildApiHandler(apiConfiguration, currentMode, new Controller(context, "commit-message-generator"))
```
Implementing

nihar-oracle replied:
Oh, ok, I just created AuthManager implementation. if that is not ok, I will use WebveiwProvider.getLastActiveInstance()

nihar-oracle commented on src/core/controller/index.ts:
```diff
@@ -376,7 +404,58 @@ export class Controller {
+			console.error("Failed to handle auth callback:", error)
+			HostProvider.window.showMessage({
+				type: ShowMessageType.ERROR,
+				message: "Failed to log in to Cline",
```
Of course, I thought the controller should be as agnostic as possible. But yes, let me update that

nihar-oracle commented on src/core/api/providers/oca.ts:
```diff
@@ -0,0 +1,262 @@
+
+	constructor(options: OcaHandlerOptions) {
+		this.options = options
+		this.options = options
```
fixing

nihar-oracle commented on webview-ui/src/components/settings/providers/OcaProvider.tsx:
```diff
@@ -0,0 +1,356 @@
+		async function tryRefresh(retry = false): Promise<boolean> {
+			try {
+				const resp = await ModelsServiceClient.refreshOcaModels(StringRequest.create({ value: baseUrl || "" }))
+				if (resp.error) throw new Error(resp.error)
```
do you run npm run lint? what warnings and what is the suggestions?

nihar-oracle commented on webview-ui/src/components/settings/providers/OcaProvider.tsx:
```diff
@@ -0,0 +1,356 @@
+				setModels(resp.models || {})
+				setError(null)
+				return true
+			} catch (err) {
```
Do you mean the error field?

Updated GitHub reply:

We renamed error (string | null) to hasError (boolean) because the UI only uses this state to toggle the recovery banner, not to render the backend error text.

What changed and why:

- The hook (useOcaModels) now tracks hasError instead of error. We set hasError=true on provider/network failure and false on success or when logging out. This matches the actual usage (visibility flag).
- The consumer destructures hasError as ocaHasError and uses it solely to show/hide the banner: {isAuthenticated && ocaHasError && ( ... )}
- The banner content is intentionally static (“Failed to refresh models. Check your session or network.”) and offers Retry / Sign in again actions. We don’t display the backend error string, so a boolean is the right type and avoids confusion/lint noise.

If we later decide to show the detailed error, we can introduce errorMessage: string | null and render it (e.g., {errorMessage || fallback}) without changing the visibility boolean.

nihar-oracle commented on webview-ui/src/components/settings/providers/OcaModelPicker.tsx:
```diff
@@ -0,0 +1,190 @@
+	onRefresh,
+}: OcaModelPickerProps) => {
+	const { handleModeFieldsChange } = useApiConfigurationHandlers()
+	const [pendingModelId, setPendingModelId] = React.useState<string | null>(null)
```
Oh, sorry, which part?

nihar-oracle commented on src/services/auth/oca/OcaAuthService.ts:
```diff
@@ -0,0 +1,226 @@
+		return this._ocaAuthState?.apiKey ?? null
+	}
+
+	async createAuthRequest(controller: Controller): Promise<ProtoString> {
```
we discussed this yesterday, right? Still needed?

nihar-oracle commented on proto/cline/models.proto:
```diff
@@ -127,6 +129,48 @@ message UpdateApiConfigurationRequest {
+ // Model info for OCA (OpenAI-compatible) models exposed by the OCA provider
+message OcaModelInfo {
+  // Maximum completion tokens per request supported by this model
+  optional int32 max_tokens = 1;
```
implementaing

nihar-oracle commented on src/services/auth/oca/OcaAuthService.ts:
```diff
@@ -0,0 +1,226 @@
+			await this._refreshInFlight
+			return
+		}
+		this._refreshInFlight = (async () => {
+			try {
+				await this.restoreRefreshTokenAndRetrieveAuthInfo(controller)
+			} finally {
+				this._refreshInFlight = null
+			}
+		})()
+		await this._refreshInFlight
```
Oh, it just prevents multiple requests in parallel, race condition

nihar-oracle commented on src/services/auth/oca/providers/OcaAuthProvider.ts:
```diff
@@ -0,0 +1,175 @@
+	private static pkceStateMap: Map<string, { code_verifier: string; nonce: string; createdAt: number; redirect_uri: string }> =
+		new Map()
+
+	protected _config: any
```
implementing

nihar-oracle commented on src/services/auth/oca/OcaAuthService.ts:
```diff
@@ -0,0 +1,226 @@
+import { getOcaConfig } from "./utils/utils"
+// import { AuthHandler } from "@/hosts/external/AuthHandler"
+
+export class OcaAuthService {
```
After the discussion yesterday, is this still needed?

celestial-vault commented on src/services/auth/oca/OcaAuthService.ts:
```diff
@@ -0,0 +1,226 @@
+	async restoreRefreshTokenAndRetrieveAuthInfo(controller: Controller): Promise<void> {
+		const provider = this.requireProvider()
+		try {
+			this._ocaAuthState = await provider.retrieveOcaAuthState(controller)
```
@nihar-oracle Wasn't seeing it in my linter. Okay

celestial-vault commented on src/integrations/git/commit-message-generator.ts:
```diff
@@ -78,7 +79,7 @@ Commit message:`
 
 		// Build the API handler
-		const apiHandler = buildApiHandler(apiConfiguration, currentMode)
+		const apiHandler = buildApiHandler(apiConfiguration, currentMode, new Controller(context, "commit-message-generator"))
```
@nihar-oracle Can you make it optional? Making a new controller here is a bit weird.

nihar-oracle replied:
I removed it, and am creating an authmanager,
This ensures that the controller is stored in a singleton and doesn't need to be passed around.
Further, this ensures that the oca code is never invoked for non oca users

celestial-vault commented on webview-ui/src/components/settings/providers/OcaProvider.tsx:
```diff
@@ -0,0 +1,356 @@
+	const debounceTimerRef = useRef<number | null>(null)
+
+	const doRefresh = useCallback(async (url: string) => {
+		const myReqId = ++reqIdRef.current
```
I have not seen ++ commonly used in React. But it's fine.

celestial-vault commented:
@kefayati Let's merge this PR first and then circle back.

celestial-vault commented:
This is already taking up a lot of time.

nihar-oracle commented on src/core/api/providers/oca.ts:
```diff
@@ -0,0 +1,261 @@
+	protected initializeClient(options: OcaHandlerOptions) {
+		return new (class OCIOpenAI extends OpenAI {
+			protected override async prepareOptions(opts: FinalRequestOptions<unknown>): Promise<void> {
+				const token = await OcaAuthService.getInstance()?.getAuthToken(options.controller)
```
I replaced this logic with AuthManger.

nihar-oracle commented on src/integrations/git/commit-message-generator.ts:
```diff
@@ -91,7 +92,7 @@ async function performCommitGeneration(context: vscode.ExtensionContext, gitDiff
 		const currentMode = "act"
 
 		// Build the API handler
-		const apiHandler = buildApiHandler(apiConfiguration, currentMode)
```
I am now using the AuthManager, so I do not need to pass in the constructor anymore

sjf reviewed (commented):
This PR is really big, and its hard for us to review as well because we don't have access to the OCA provider. 
We can get it submitted faster if you split into smaller parts, then we can review faster and start getting it submitted incrementally.

nihar-oracle commented:
> Hi, we split the front end and the backend, as we'd discussed. Access to oca will be part of the second pr Happy to make it even smaller 😅. But not sure how

Hi, we split the front end and the backend, as we'd discussed.
Access to oca will be part of the second pr

Happy to make it even smaller 😅.

But not sure how

nihar-oracle committed:
fix: Removing AuthManager, simplifyng auth service initialization

nihar-oracle commented:
Discussed with @celestial-vault offline.
1. Remove AuthManager
2. Separate get instance from initialize on OcaAuthService

nihar-oracle commented:
> Discussed with @celestial-vault offline.
> 
> 1. Remove AuthManager
> 2. Separate get instance from initialize on OcaAuthService

Pushed reply

sjf commented on src/extension.ts:
```diff
@@ -57,6 +58,13 @@ export async function activate(context: vscode.ExtensionContext) {
 
+	// Ensure auth services are initialized with the controller
+	{
+		const controller = sidebarWebview.controller
```
The shared set-up code for VSCode, JetBrains, etc. is in common.ts, can you move the initialization there, so then it will work on both platforms.

nihar-oracle replied:
I will move the Auth initialization to common.ts

nihar-oracle replied:
Thanks, @saoudrizwan  and @sjf, for your feedback.

After further review and discussion, I realized that moving the OCAAuthService initialization into common.ts isn’t straightforward at the moment, since common.ts doesn’t have access to the controller (stateManager) required for initializing OCAAuthService. The controller is currently only available within controller/index.ts.

Therefore, I’ll keep the initialization in controller/index.ts for now and remove it from extension.ts to avoid duplication. If we move to a setup where the controller is accessible in common.ts in the future, we can revisit this refactor for better cross-platform consistency.

Please let me know if you have additional suggestions.

sjf replied:
Thanks Nihar, I am happy to discuss this more, but for a first iteration I think your approach is fine.

nihar-oracle commented on src/extension.ts:
```diff
@@ -57,6 +58,13 @@ export async function activate(context: vscode.ExtensionContext) {
+	// Ensure auth services are initialized with the controller
+	{
+		const controller = sidebarWebview.controller
+		AuthService.getInstance(controller)
```
Thanks, @saoudrizwan and @sjf, for your feedback.

After further review and discussion, I realized that moving the OCAAuthService initialization into common.ts isn’t straightforward at the moment, since common.ts doesn’t have access to the controller (stateManager) required for initializing OCAAuthService. The controller is currently only available within controller/index.ts.

Therefore, I’ll keep the initialization in controller/index.ts for now and remove it from extension.ts to avoid duplication. If we move to a setup where the controller is accessible in common.ts in the future, we can revisit this refactor for better cross-platform consistency.

Please let me know if you have additional suggestions.

sjf commented:
The PR looks good to me, Nihar has made all the changes we've asked for.  I've tested the PR locally and I don't see any issues with the existing Cline auth. 

https://github.com/user-attachments/assets/54efb47f-5840-4506-bfc3-f19458106559

celestial-vault reviewed (approved):
Looks good 👍

shirazhussain-oracle commented on src/core/controller/index.ts:
```diff
@@ -169,6 +172,23 @@ export class Controller {
+	// Oca Auth methods
+	async handleOcaSignOut() {
+		try {
+			this.ocaAuthService.handleDeauth()
```
Should we do `await` here?

shirazhussain-oracle commented on src/services/auth/oca/providers/OcaAuthProvider.ts:
```diff
@@ -0,0 +1,184 @@
+			})
+			// Step 1: Nonce validation
+			const idToken = tokenResponse.data.id_token
+			if (idToken) {
```
what happens when idToken is falsy?

shirazhussain-oracle commented on src/core/controller/models/refreshOcaModels.ts:
```diff
@@ -0,0 +1,135 @@
+	}
+	const models: Record<string, OcaModelInfo> = {}
+	let defaultModelId: string | undefined
+	const ocaAccessToken = await OcaAuthService.getInstance().getAuthToken()
```
Check for missing token here and fail fast with informative error via HostProvider.window.showMessage

shirazhussain-oracle commented on src/services/auth/oca/OcaAuthService.ts:
```diff
@@ -0,0 +1,248 @@
+			throw new Error("IDCS URI is not configured")
+		}
+		// Start the auth handler
+		const callbackUrl = `${await AuthHandler.getInstance().getCallbackUrl()}\auth\oca`
```
Should it be forward slashes `/auth/oca` ?
